File: c_toolkit/src/automation.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../include/bash_toolkit.h"

int health_monitor(int interval, int cpu_threshold, int mem_threshold, int disk_threshold, int continuous, const char *output_file, int dry_run) {
    char *args[20];
    int i = 0;

    char int_str[16]; snprintf(int_str, sizeof(int_str), "%d", interval);
    args[i++] = "-i"; args[i++] = int_str;

    char cpu_str[16]; snprintf(cpu_str, sizeof(cpu_str), "%d", cpu_threshold);
    args[i++] = "-c"; args[i++] = cpu_str;

    char mem_str[16]; snprintf(mem_str, sizeof(mem_str), "%d", mem_threshold);
    args[i++] = "-m"; args[i++] = mem_str;

    char disk_str[16]; snprintf(disk_str, sizeof(disk_str), "%d", disk_threshold);
    args[i++] = "-d"; args[i++] = disk_str;

    if (continuous) args[i++] = "-C";
    if (output_file) { args[i++] = "-o"; args[i++] = (char *)output_file; }

    args[i] = NULL;
    return run_script("bash_toolkit/scripts/automation/health_monitor.sh", args, dry_run);
}

int auto_backup(const char *source, const char *backup_dir, int retention_days, const char *exclude_patterns[], int num_excludes, int no_compression, const char *remote_host, int dry_run) {
    char *args[64];
    int i = 0;

    args[i++] = "-s"; args[i++] = (char *)source;
    args[i++] = "-b"; args[i++] = (char *)backup_dir;

    char ret_str[16]; snprintf(ret_str, sizeof(ret_str), "%d", retention_days);
    args[i++] = "-r"; args[i++] = ret_str;

    if (exclude_patterns) {
        for (int j = 0; j < num_excludes; j++) {
            if (i < 60) {
                args[i++] = "-e"; args[i++] = (char *)exclude_patterns[j];
            }
        }
    }

    if (no_compression) args[i++] = "-n";
    if (remote_host) { args[i++] = "-R"; args[i++] = (char *)remote_host; }

    args[i] = NULL;
    return run_script("bash_toolkit/scripts/automation/auto_backup.sh", args, dry_run);
}

int service_watchdog(const char *services[], int num_services, int interval, int max_restarts, const char *email, int run_once, int dry_run) {
    char *args[64];
    int i = 0;

    if (services) {
        for (int j = 0; j < num_services; j++) {
            if (i < 50) {
                args[i++] = "-s"; args[i++] = (char *)services[j];
            }
        }
    }

    char int_str[16]; snprintf(int_str, sizeof(int_str), "%d", interval);
    args[i++] = "-i"; args[i++] = int_str;

    char max_str[16]; snprintf(max_str, sizeof(max_str), "%d", max_restarts);
    args[i++] = "-m"; args[i++] = max_str;

    if (email) { args[i++] = "-e"; args[i++] = (char *)email; }
    if (run_once) args[i++] = "-1";

    args[i] = NULL;
    return run_script("bash_toolkit/scripts/automation/service_watchdog.sh", args, dry_run);
}

int deploy_helper(const char *project_dir, const char *deploy_type, const char *branch, int run_tests, int no_backup, const char *restart_services[], int num_restart_services, int dry_run) {
    char *args[64];
    int i = 0;

    args[i++] = "-p"; args[i++] = (char *)project_dir;
    args[i++] = "-t"; args[i++] = (char *)deploy_type;
    args[i++] = "-b"; args[i++] = (char *)branch;

    if (run_tests) args[i++] = "-T";
    if (no_backup) args[i++] = "-n";

    if (restart_services) {
        for (int j = 0; j < num_restart_services; j++) {
            if (i < 60) {
                args[i++] = "-r"; args[i++] = (char *)restart_services[j];
            }
        }
    }

    args[i] = NULL;
    return run_script("bash_toolkit/scripts/automation/deploy_helper.sh", args, dry_run);
}

int batch_processor(const char *operation, const char *directory, const char *pattern, const char *new_extension, const char *prefix, const char *suffix, const char *find_text, const char *replace_text, const char *compress_type, int dry_run_script, int dry_run) {
    char *args[32];
    int i = 0;

    args[i++] = "-o"; args[i++] = (char *)operation;
    args[i++] = "-d"; args[i++] = (char *)directory;
    args[i++] = "-p"; args[i++] = (char *)pattern;
    args[i++] = "-c"; args[i++] = (char *)compress_type;

    if (new_extension) { args[i++] = "-e"; args[i++] = (char *)new_extension; }
    if (prefix) { args[i++] = "-P"; args[i++] = (char *)prefix; }
    if (suffix) { args[i++] = "-S"; args[i++] = (char *)suffix; }
    if (find_text) { args[i++] = "-f"; args[i++] = (char *)find_text; }
    if (replace_text) { args[i++] = "-r"; args[i++] = (char *)replace_text; }
    if (dry_run_script) args[i++] = "-n";

    args[i] = NULL;
    return run_script("bash_toolkit/scripts/automation/batch_processor.sh", args, dry_run);
}

int api_test(const char *url, const char *method, const char *headers[], int num_headers, const char *data, int expected_status, int timeout, const char *test_file, const char *output_file, int verbose, int no_follow_redirects, int dry_run) {
    char *args[64];
    int i = 0;

    if (test_file) {
        args[i++] = "-f"; args[i++] = (char *)test_file;
    } else {
        if (!url) return -1;
        args[i++] = "-u"; args[i++] = (char *)url;
        args[i++] = "-m"; args[i++] = (char *)method;

        char status_str[16]; snprintf(status_str, sizeof(status_str), "%d", expected_status);
        args[i++] = "-s"; args[i++] = status_str;

        char timeout_str[16]; snprintf(timeout_str, sizeof(timeout_str), "%d", timeout);
        args[i++] = "-t"; args[i++] = timeout_str;

        if (headers) {
            for (int j = 0; j < num_headers; j++) {
                if (i < 50) {
                    args[i++] = "-H"; args[i++] = (char *)headers[j];
                }
            }
        }

        if (data) { args[i++] = "-d"; args[i++] = (char *)data; }
    }

    if (output_file) { args[i++] = "-o"; args[i++] = (char *)output_file; }
    if (verbose) args[i++] = "-v";
    if (no_follow_redirects) args[i++] = "-n";

    args[i] = NULL;
    return run_script("bash_toolkit/scripts/automation/api_tester.sh", args, dry_run);
}

File: bash_toolkit/automation.py
from .utils import run_script

def health_monitor(interval=5, cpu_threshold=80, mem_threshold=80, disk_threshold=80, continuous=False, output_file=None, dry_run=False):
    """
    Wrapper for bash_toolkit/scripts/automation/health_monitor.sh
    """
    args = ["-i", interval, "-c", cpu_threshold, "-m", mem_threshold, "-d", disk_threshold]

    if continuous:
        args.append("-C")

    if output_file:
        args.extend(["-o", output_file])

    return run_script("bash_toolkit/scripts/automation/health_monitor.sh", args, stream=continuous, dry_run=dry_run)

def auto_backup(source, backup_dir="./backups", retention_days=7, exclude_patterns=None, no_compression=False, remote_host=None, dry_run=False):
    """
    Wrapper for bash_toolkit/scripts/automation/auto_backup.sh
    """
    args = ["-s", source, "-b", backup_dir, "-r", retention_days]

    if exclude_patterns:
        for pattern in exclude_patterns:
            args.extend(["-e", pattern])

    if no_compression:
        args.append("-n")

    if remote_host:
        args.extend(["-R", remote_host])

    return run_script("bash_toolkit/scripts/automation/auto_backup.sh", args, dry_run=dry_run)

def service_watchdog(services, interval=30, max_restarts=3, email=None, run_once=False, dry_run=False):
    """
    Wrapper for bash_toolkit/scripts/automation/service_watchdog.sh
    """
    args = ["-i", interval, "-m", max_restarts]

    if isinstance(services, list):
        for service in services:
            args.extend(["-s", service])
    else:
        args.extend(["-s", services])

    if email:
        args.extend(["-e", email])

    if run_once:
        args.append("-1")

    stream = not run_once
    return run_script("bash_toolkit/scripts/automation/service_watchdog.sh", args, stream=stream, dry_run=dry_run)

def deploy_helper(project_dir, deploy_type="web", branch="main", run_tests=False, no_backup=False, restart_services=None, dry_run=False):
    """
    Wrapper for bash_toolkit/scripts/automation/deploy_helper.sh
    """
    args = ["-p", project_dir, "-t", deploy_type, "-b", branch]

    if run_tests:
        args.append("-T")

    if no_backup:
        args.append("-n")

    if restart_services:
        if isinstance(restart_services, list):
            for service in restart_services:
                args.extend(["-r", service])
        else:
            args.extend(["-r", restart_services])

    return run_script("bash_toolkit/scripts/automation/deploy_helper.sh", args, dry_run=dry_run)

def batch_processor(operation="list", directory=".", pattern="*", new_extension=None, prefix=None, suffix=None, find_text=None, replace_text=None, compress_type="gz", dry_run_script=False, dry_run=False):
    """
    Wrapper for bash_toolkit/scripts/automation/batch_processor.sh
    """
    args = ["-o", operation, "-d", directory, "-p", pattern, "-c", compress_type]

    if new_extension:
        args.extend(["-e", new_extension])

    if prefix:
        args.extend(["-P", prefix])

    if suffix:
        args.extend(["-S", suffix])

    if find_text:
        args.extend(["-f", find_text])

    if replace_text:
        args.extend(["-r", replace_text])

    if dry_run_script:
        args.append("-n")

    return run_script("bash_toolkit/scripts/automation/batch_processor.sh", args, dry_run=dry_run)

def api_test(url=None, method="GET", headers=None, data=None, expected_status=200, timeout=30, test_file=None, output_file=None, verbose=False, no_follow_redirects=False, dry_run=False):
    """
    Wrapper for bash_toolkit/scripts/automation/api_tester.sh
    """
    args = []

    if test_file:
        args.extend(["-f", test_file])
    else:
        if url is None:
             raise ValueError("URL is required unless test_file is provided")
        args.extend(["-u", url, "-m", method, "-s", expected_status, "-t", timeout])

        if headers:
            for header in headers:
                args.extend(["-H", header])

        if data:
            args.extend(["-d", data])

    if output_file:
        args.extend(["-o", output_file])

    if verbose:
        args.append("-v")

    if no_follow_redirects:
        args.append("-n")

    return run_script("bash_toolkit/scripts/automation/api_tester.sh", args, dry_run=dry_run)
